client/src/index.js

```diff
@@ -1,30 +1,143 @@
+    }
+  })()
+}
+
+function persistReducer(state = { }, action) {
+  switch (action.type) {
+    case 'LOAD_PERSIST':
+      return action.state
+
+    case 'SET_PERSIST':
+      return { ...state, ...action.data }
+
+    default:
+      return state
+  }
+}
+
+const reducer = combineReducers({
+  persist: persistReducer,
+})
+
+const store = createStore(reducer)
+
+const savePersist = setupPersist((state) => {
+  store.dispatch({ type: 'LOAD_PERSIST', state })
+})
+store.subscribe(() => {
+  savePersist(store.getState().persist || { })
+})
+
+function versionString(version) {
+  const shortHash = version.gitHash.substring(0, 8)
+  return `${shortHash} (${version.gitBranch})`
+}
+
+class NameInput extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = { name: null }
+
+    this.save = this.save.bind(this)
+    this.change = this.change.bind(this)
+  }
+
+  change(e) {
+    this.setState({ name: e.target.value })
+  }
+
+  save() {
+    const name = this.state.name !== null ? this.state.name : this.props.name
+    this.props.onSave(name)
+  }
+
+  render() {
+    const { name: stateName } = this.state
+    const { name: propName } = this.props
+    const name = stateName !== null ? stateName : propName
+    return <>
+      <input type="text" value={name} onChange={this.change} />
+      <button type="button" onClick={this.save}>Save</button>
+    </>
+  }
+}
+
+const PersistNameInput = connect(
+  state => ({ name: state.persist.name || '' }),
+  dispatch => ({
+    onSave: (name) => {
+      dispatch({
+        type: 'SET_PERSIST',
+        data: { name },
+      })
+    },
+  }),
+)(NameInput)
+
+function NameDisplay({ name, prefix }) {
+  return <>{prefix}{name}</>
+}
+
+const PersistNameDisplay = connect(
+  state => ({ name: state.persist.name || '' }),
+)(NameDisplay)
+
+class ApiVersionInfo extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = { }
   }
 
   async componentDidMount() {
     const response = await fetch('/api/version')
     const version = await response.json()
-    const shortHash = version.gitHash.substring(0, 8)
-    const versionString = `${shortHash} (${version.gitBranch})`
-    this.setState({ versionString })
+    this.setState(version)
   }
 
   render() {
-    const { versionString } = this.state
+    const { ...state } = this.state
     return (
-      <div>Version: {versionString}</div>
+      <div>API Version: {state.gitHash ? versionString(state) : null}</div>
     )
   }
 }
 
+function ClientVersionInfo({ version }) {
+  return (
+    <div>Client Version: {versionString(version)}</div>
+  )
+}
+
 ReactDOM.render(
-  <VersionInfo />,
+  <Provider store={store}>
+    <>
+      <div>
+        <ApiVersionInfo />
+        <ClientVersionInfo version={clientVersion} />
+      </div>
+      <div>
+        <PersistNameInput />
+      </div>
+      <div>
+        <PersistNameDisplay prefix="Persisted name: " />
+      </div>
```

Comment:
Are these `div`s necessary? They don't seem to serve purpose.

Reply:
They are used to layout the names on top of each other

Reply:
temporary I assume

Reply:
@bqqbarbhg better to remove the divs and control the rendering box in css directly. So with display: block in this case.

Note for later: we could also discuss to opt for flexbox (as it would be good for (especially) frontend guys to learn)

Reply:
Yea all the UI here is *extremely* temporary, just for trying out the persistent state.

@HenrikHelenius I vote for flexbox, a good way to keep your sanity with css